### PR TITLE
Dev calculator agent

### PR DESCRIPTION
**Pull request by szhang012**
This branch contains all agent implementations related to the Mackay Calculator:
1. Mackay Calculator Agent
2. Mackay Data Agent
3. API Agent.

### CODE REVIEW

File: Agents/MackayCalculatorAgent/api/ModelTwerkHandler.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Data Handler to twerk the handler to back to net zero after an specified action on a specified lever, the new plot charts dta are returned
+# This router is still in development and not used for deployment
+# ===============================================================================
+from flask_restful import Api, Resource, reqparse,request
+from mackay.calculator_model import CalculatorModel
+from flask import jsonify
+
+calculator = CalculatorModel()
+class ModelTwerkHandler(Resource):
```
Comment (sm453) [inline Agents/MackayCalculatorAgent/api/ModelTwerkHandler.py:9]: Are you sure you are using the word "twerk" correctly? As far as I'm aware it means something very _very_ different. Please google the meaning of it, to be sure that's what you actually mean. Perhaps you mean "tweak"?

File: Agents/MackayCalculatorAgent/frontend/src/App.test.js
Comment (sm453) [inline Agents/MackayCalculatorAgent/frontend/src/App.test.js:1]: Is this file needed?

File: Agents/MackayCalculatorAgent/mackay/calculator_model.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    #validate level config are in crrect range (1-4)
+    def validateLevel(self, levels):
+        #print('validated that post levels are integers')
+        for l in levels:
+            if l not in VALID_LEVELS:
+                return False
+        return True
+
+    #twerk till netzero
+    def twerkTillZero(self, action, pos_cate, current_config):
```
Comment (sm453) [inline Agents/MackayCalculatorAgent/mackay/calculator_model.py:41]: Same comment as above.

File: Agents/MackayCalculatorAgent/readme.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+python >=3.10, java jdk>=11, nodeJS>=18
+
+###Setup
+
+
+1. Install required python packages.
+```shell
+pip install -r requirements.txt
+```
+2. Download Mackay Calculator Excel model and mapping csvs (```controls.csv```, ```output.csv```,```single_values.csv```) into ```model/```.
```
Comment (sm453) [inline Agents/MackayCalculatorAgent/readme.md:49]: Please add a hint where to get these from. A good location could be `dropbox/IRP3 CAPRICORN shared folder/_TWA_Shared_Data/`.

File: Agents/MackayCalculatorAgent/test/unit/webapp/client.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+@pytest.fixture
+def client():
+    """Configures the app for testing
+
+    Sets app config variable ``TESTING`` to ``True``
+
+    :return: App for testing
+    """
+
+    #app.config['TESTING'] = True
```
Comment (sm453) [inline Agents/MackayCalculatorAgent/test/unit/webapp/client.py:13]: This is highly confusing. This file is in a test folder, so one would assume it is for testing, and the comment strongly suggests that `TESTING` should be set to `True`, yet it is commented out. Please resolve this.

File: Agents/MackayCalculatorAgent/test/unit/webapp/test_root.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+@pytest.fixture
+def client():
+    """Configures the app for testing
+
+    Sets app config variable ``TESTING`` to ``True``
+
+    :return: App for testing
+    """
+
+    #app.config['TESTING'] = True
```
Comment (sm453) [inline Agents/MackayCalculatorAgent/test/unit/webapp/test_root.py:15]: Same comment as above.

File: Agents/MackayDataAgent/README.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+2. PostgresSQL.
+3. [Forecasting Agent]
+4. [API Agent]
+#2. Agent Design
+
+# 2.1 Mackay Calculator Agents
+The Singapore MacKay Carbon Calculator provides a model of the Singapore energy system that allows you to explore pathways to decarbonisation. The scenario simulation is based on an underling Mackay Calculator Model. We implement it in the TWA agent framework with two agents.
+1) Mackay Calculator Agent: The agent that manages the web-interface and controlling API over the Calculator model.
+2) Mackay Data Agent: The agent that manages the update of the derived Timeseries (TS) datapoints in TWA that is required as the input to the Calculator Model. The required input TS data are derived under DIF with a chain of [Forecasting Agent] and [API Agent].
+
```
Comment (sm453) [inline Agents/MackayDataAgent/README.md:80]: So, can these two agents be run independently? Or is one dependent on the other? Please clarify. It would be helpful to state this explicitly.

File: Agents/APIAgent/README.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+Some source-data-to-target-Timeseries mapping may need algebra operations. For examples, the raw data contains 4 set of sub-category quantities and the target quantity is the sum of them. Such operations can be declared as a Named Individual of type `rmlts:Algebra`. And a corresponding implementation should be put in the [calculations module]. We have implemented one example: `rmlts:Sum`. An API meta-data example can be found in [test_triples] folder: [api_pvcapacity.ttl].
+
+
+
+
+#3. Using the Agent
+
+A local running container can be initiated via:
+```
+docker compose -f "docker-compose-test-dockerised.yml" up -d --build
```
Comment (sm453) [inline Agents/APIAgent/README.md:116]: The fact that it's "dockerised" goes without saying, given that it's a docker compose file. To call it "test" is confusing though, as this is normally reserved for unit/regression/integration tests. I suspect that this is meant to be used not in that sense, so perhaps it should be called `docker-compose.yml`?

**Review by sm453 — CHANGES_REQUESTED** [review]

**Comment by sm453** [comment]
If the suggested changes to the ontology are not immediately feasible, then please at the very least add a to-do list with all the mentioned items to the relevant readme file.

**Comment by szhang012** [comment]
Dear Seb,

Greetings. I believe the changes are very feasible. I will address them today.


Regards
Shaocong


________________________________
发件人: Sebastian Mosbach ***@***.***>
发送时间: 2024年4月16日 14:02
收件人: cambridge-cares/TheWorldAvatar ***@***.***>
抄送: #ZHANG SHAOCONG# ***@***.***>; Author ***@***.***>
主题: Re: [cambridge-cares/TheWorldAvatar] Dev calculator agent (PR #1154)


If the suggested changes to the ontology are not immediately feasible, then please at the very least add a to-do list with all the mentioned items to the relevant readme file.

―

File: Agents/APIAgent/docker-compose-test-dockerised.yml [hunk 8]
```diff
@@ -0,0 +1,21 @@
+version: '3.8'
+
+services:
+  api_agent:
+    image: ghcr.io/cambridge-cares/api_agent:1.0.0-SNAPSHOT
```
[the remark this answers is not in the document]
Comment (szhang012) [reply Agents/APIAgent/docker-compose-test-dockerised.yml]: This should be fixed in the latest commit.

File: Agents/APIAgent/tbox_dev/rmlts.owl [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    ///////////////////////////////////////////////////////////////////////////////////////
+     -->
+    <owl:DatatypeProperty rdf:about="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#time_iterator">
+        <rdfs:domain rdf:resource="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#TimeSeriesMap"/>
+    </owl:DatatypeProperty>
+
+
+    <owl:DatatypeProperty rdf:about="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#value_iterator">
+        <rdfs:domain rdf:resource="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#TimeSeriesMap"/>
+    </owl:DatatypeProperty>
```
[the remark this answers is not in the document]
Comment (szhang012) [reply Agents/APIAgent/tbox_dev/rmlts.owl]: I have modified it. However since this file is going to the JPS_ONTOLOGY folder, I suppose I should delete it in this repository and later commit to the JPS_ONTOLOGY in a separate branch.

File: Agents/APIAgent/tbox_dev/rmlts.owl [hunk 10]
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0"?>
+<rdf:RDF xmlns="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#"
+     xml:base="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl"
+     xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#"
+     xmlns:owl="http://www.w3.org/2002/07/owl#"
+     xmlns:xml="http://www.w3.org/XML/1998/namespace"
+     xmlns:xsd="http://www.w3.org/2001/XMLSchema#"
+     xmlns:rdfs="http://www.w3.org/2000/01/rdf-schema#"
+     >
+    <owl:Ontology rdf:about="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl"/>
```
[the remark this answers is not in the document]
Comment (szhang012) [reply Agents/APIAgent/tbox_dev/rmlts.owl]: This has been addressed in the latest commit.

File: Agents/APIAgent/tbox_dev/rmlts.owl [hunk 11]
```diff
@@ -0,0 +1,75 @@
+<?xml version="1.0"?>
+<rdf:RDF xmlns="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#"
```
[the remark this answers is not in the document]
Comment (szhang012) [reply Agents/APIAgent/tbox_dev/rmlts.owl]: I change the RML4TS references in the readme all to lowercases, as in the new commit.

File: Agents/APIAgent/tbox_dev/rmlts.owl [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    </owl:Class>
+    <!--
+    ///////////////////////////////////////////////////////////////////////////////////////
+    //
+    // Individuals
+    //
+    ///////////////////////////////////////////////////////////////////////////////////////
+     -->
+    <owl:NamedIndividual rdf:about="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#Sum">
+        <rdf:type rdf:resource="http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#Algebra"/>
```
[the remark this answers is not in the document]
Comment (szhang012) [reply Agents/APIAgent/tbox_dev/rmlts.owl]: I change Algebra to MapFunction, if that is better... As for enumeration, my initial thought was closer to an enumeration in coding, as a list of predefined concepts (in this case, mapping functions). However for the Mackay data I only need one - the sum.

File: Agents/MackayCalculatorAgent/api/ModelTwerkHandler.py (hunk 1 above)
Comment (szhang012) [reply Agents/MackayCalculatorAgent/api/ModelTwerkHandler.py]: Nope, that was definitely not what I mean. Yes I mean 'tweak'.

File: Agents/MackayCalculatorAgent/frontend/src/App.test.js
Comment (szhang012) [reply Agents/MackayCalculatorAgent/frontend/src/App.test.js:1]: Deleted.

File: Agents/MackayCalculatorAgent/mackay/calculator_model.py (hunk 2 above)
Comment (szhang012) [reply Agents/MackayCalculatorAgent/mackay/calculator_model.py]: Modified in commit and deleted in mind.

File: Agents/MackayCalculatorAgent/readme.md (hunk 3 above)
Comment (szhang012) [reply Agents/MackayCalculatorAgent/readme.md]: I have put the files in Dropbox and add a link to README.

File: Agents/MackayCalculatorAgent/test/unit/webapp/client.py (hunk 4 above)
Comment (szhang012) [reply Agents/MackayCalculatorAgent/test/unit/webapp/client.py]: Yes this comment-out is uncommented out.

Comment (szhang012) [reply Agents/MackayCalculatorAgent/test/unit/webapp/client.py]: I find this is an unused file I forgot to delete and it is deleted in the new commit.

File: Agents/MackayCalculatorAgent/test/unit/webapp/test_root.py (hunk 5 above)
Comment (szhang012) [reply Agents/MackayCalculatorAgent/test/unit/webapp/test_root.py]: The comment-out is uncommented out.

File: Agents/MackayDataAgent/README.md (hunk 6 above)
Comment (szhang012) [reply Agents/MackayDataAgent/README.md:83]: This is clarified in READMES of both agents.

File: Agents/APIAgent/README.md (hunk 7 above)
Comment (szhang012) [reply Agents/APIAgent/README.md]: Changed to 'docker-compose'.

File: Agents/APIAgent/tbox_dev/rmlts.owl
[the remark this answers is not in the document]
Comment (sm453) [reply Agents/APIAgent/tbox_dev/rmlts.owl:1]: I note you have simply deleted this file, rather than move it to a better place. Is that what you intended? I believe it would be preferable to move the file as part of this pull request, rather than doing that on a separate branch.

File: Agents/APIAgent/kg_access/apisource_kg_client.py [hunk 13]
```diff
@@ -10,7 +10,7 @@
 INSERT_STR = "INSERT DATA"
 
 META_PREFIX = '''
-prefix rml4ts: <http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#>
+prefix rml4ts: <https://www.theworldavatar.com/kg/rml4ts.owl#>
```
Comment (sm453) [inline Agents/APIAgent/kg_access/apisource_kg_client.py:13]: You seem to have missed two occurrences of "4"...

File: Agents/APIAgent/data_classes/iris.py [hunk 14]
```diff
@@ -9,7 +9,7 @@
 XSD = "http://www.w3.org/2001/XMLSchema#"
 TIME = "http://www.w3.org/2006/time#"
 TS = "https://www.theworldavatar.com/kg/ontotimeseries/"
-RML4TS = "http://www.theworldavatar.com/ontology/rml4ts/rml4ts.owl#"
+RML4TS = "https://www.theworldavatar.com/kg/rml4ts.owl#"
```
Comment (sm453) [inline Agents/APIAgent/data_classes/iris.py:12]: You seem to have missed two occurrences of "4"...

File: Agents/APIAgent/README.md [hunk 15]
```diff
@@ -101,7 +101,7 @@ Once this is done, the API Agent automatically manages the instantiation and all
 
 ##2.2 Extra Options
 ###2.2.1 API with dynamically generated URL
-It is common for external data API to use dynamically generated access URL. For this purpose, we define a new rdf property in `RMLTS.OWL`: `rmlts:hasDynamicTarget`. Find in [test_triples] folder an example to define an API source with dynamically generated URL: [api_temperature.ttl].
+It is common for external data API to use dynamically generated access URL. For this purpose, we define a new rdf property in `rmlts.owl`: `rmlts:hasDynamicTarget`. Find in [test_triples] folder an example to define an API source with dynamically generated URL: [api_temperature.ttl].
  
 ###2.2.2 API-to-TS mapping with Algebra
 Some source-data-to-target-Timeseries mapping may need algebra operations. For examples, the raw data contains 4 set of sub-category quantities and the target quantity is the sum of them. Such operations can be declared as a Named Individual of type `rmlts:Algebra`. And a corresponding implementation should be put in the [calculations module]. We have implemented one example: `rmlts:Sum`. An API meta-data example can be found in [test_triples] folder: [api_pvcapacity.ttl].
```
Comment (sm453) [inline Agents/APIAgent/README.md:107]: Does the "Algebra" bit need updating?

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Agents/APIAgent/tbox_dev/rmlts.owl (hunk 9 above)
[the remark this answers is not in the document]
Comment (sm453) [reply Agents/APIAgent/tbox_dev/rmlts.owl]: As commented, I believe it would be preferable to simply move the file as part of this pull request.

File: Agents/APIAgent/tbox_dev/rmlts.owl
[the remark this answers is not in the document]
Comment (szhang012) [reply Agents/APIAgent/tbox_dev/rmlts.owl:1]: Yes, I intended to upload it by creating a separate branch which was why I deleted it. If it is preferable to move it in this request then I will upload it in the change commit.

File: Agents/APIAgent/tbox_dev/rmlts.owl (hunk 9 above)
[the remark this answers is not in the document]
Comment (szhang012) [reply Agents/APIAgent/tbox_dev/rmlts.owl]: I have checked the TBOX generator, however, I have questions concerning that it does not seem to contain a way to define NamedIndividual. Does it mean we are not using this definition anymore? Either way, I put an manually edited updated version tentatively in the latest pull request.

File: Agents/APIAgent/README.md (hunk 15 above)
Comment (szhang012) [reply Agents/APIAgent/README.md]: This has been addressed in the latest change commit.

File: Agents/APIAgent/data_classes/iris.py (hunk 14 above)
Comment (szhang012) [reply Agents/APIAgent/data_classes/iris.py]: This has been addressed in the latest change commit. The name is unified to 'rml4ts', as 'RML for Timeseries', I errorneously put 'rmlts' in the file names and READMES.

File: Agents/APIAgent/tbox_dev/rmlts.owl (hunk 9 above)
[the remark this answers is not in the document]
Comment (sm453) [reply Agents/APIAgent/tbox_dev/rmlts.owl]: The current design of the TBoxGenerator indeed only supports TBox statements, i.e. concepts/properties, but not instances, i.e. named individuals.

Since the ontology is tiny, it makes most sense to take no further action in this case.

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Agents/APIAgent/resources/APIAgent.properties
Comment (sm453) [inline Agents/APIAgent/resources/APIAgent.properties:1]: This file seems to duplicate information in the `Agents/APIAgent/agent.env.*` files. Are they, and the present file, all required?

If the `agent.env.*` files are no longer required, please delete them, and remember to update both the `Dockerfile` and the `docker-compose.yml` file, which all reference them.

The `README.md` only refers to the `agent.env.*` files, but not the present file. Can you please add/change the explanation in the readme file?

**Review by sm453 — CHANGES_REQUESTED** [review]

File: Agents/APIAgent/resources/APIAgent.properties
Comment (szhang012) [reply Agents/APIAgent/resources/APIAgent.properties:1]: I have also put the below explanation in my latest commit of README: `agent.env.run` is used for running the agent in Docker. `agent.env.test` is used in pytest. The need of two separate files is due to 'localhost' should be referred to as 'host.docker.internal' from inside Docker.
